### PR TITLE
Switch Dependency Management to pyproject.toml + Bonus Changes

### LINKED ISSUE #1395: CONTRIBUTING.md should recommend python's breakpoint() instead of "import pdb; pdb.set_trace()"
Does the same thing but more convenient. Did not exist when I learned about pdb, but now everyone has a new enough python for using `breakpoint()`.

### LINKED ISSUE #1403: Add the `pytest-randomly` Pytest Plugin
As I thought through https://github.com/Akuli/porcupine/issues/1401 it occurred to me that there might be some incidental dependence on the order of tests. I looked into what solutions existed for this and [`pytest-randomly`](https://pypi.org/project/pytest-randomly/) seems to be the most popular one. It also supports some nice features for selecting a specific random seed which may help with other flaky tests in time.

To implement, literally all that needs done is to install it and it uses it by default.

To disable once installed: `pytest -p no:randomly`.

I have ran the test suite several times locally using it and everything no tests fail, which is great. I notice no difference in test collection time.

### PR DESCRIPTION
**Pull request by benjamin-kirkbride**
Fixes https://github.com/Akuli/porcupine/issues/1381
Fixes #1403
Fixes #1395
Fixes https://github.com/Akuli/porcupine/issues/1317

Added some docs related to https://github.com/Akuli/porcupine/issues/1219

### CODE REVIEW

**Comment by benjamin-kirkbride** [comment]
okay, I'm done!

**Comment by benjamin-kirkbride** [comment]
Not sure what is wrong with Windows

**Review by Akuli — COMMENTED** [review]
Most of my comments are about dependency management. We can split dependency stuff to a separate PR, or do it all here, however you want.

File: .github/workflows/autofix.yml [hunk 1]
```diff
@@ -22,15 +22,15 @@ jobs:
...
-    - run: pip install -r requirements-dev.txt
+    - run: pip install ".[dev]"
     - name: Gather a list of Python files in the pull request branch
       run: |
         (cd pr && git ls-files) | grep -E '\.(py|pyw)$' | sed s:^:pr/: | tee filelist.txt
     - run: python3 -m pycln --all --disable-all-dunder-policy $(cat filelist.txt)
     - run: python3 -m pyupgrade $(cat filelist.txt) --keep-runtime-typing --py38-plus --exit-zero-even-if-changed
     - run: python3 -m black $(cat filelist.txt)
-    - run: python3 -m isort $(cat filelist.txt)
+    - run: python3 -m ruff $(cat filelist.txt)
```
Comment (Akuli) [inline .github/workflows/autofix.yml:32]: Can we split ruff stuff into a separate PR? I want it to be easily revertible, especially the autofix stuff, because the autofix isn't exactly very easy to test before merging.

File: README.md [hunk 2]
```diff
@@ -70,19 +71,27 @@ Install Python 3.8 or newer with pip and tkinter somehow.
...
 and check "Show Porcupine in the desktop menu system".
 This makes Porcupine show up in the menu just like any other application.
 
+To update Porcupine to the latest version:
+
+```bash
+pipx install --force git+https://github.com/Akuli/porcupine.git
+```
+
+This may nuke your configs, so be sure to back them up if you care about them.
```
[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply README.md:91]: I think the configs end up in the venv that is automatically managed by `pipx`? I guess I should probably test it and say for sure.

File: README.md [hunk 3]
```diff
@@ -70,19 +71,27 @@ Install Python 3.8 or newer with pip and tkinter somehow.
...
 Then run these commands:
 
-    python3 -m venv porcupine-venv
-    source porcupine-venv/bin/activate
-    pip install wheel
-    pip install https://github.com/Akuli/porcupine/archive/v2023.06.27.zip
-    porcu
+```bash
+python3 -m pip install -u pipx # if you don't have `pipx` already
+pipx install git+https://github.com/Akuli/porcupine.git
```
[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply README.md:76]: > Why do we use pipx for "Other Linux distributions" but not for "Debian-based Linux distributions"?

I didn't notice the Debian section :sweat_smile: 

> I thought the whole point of pipx is that you just tell it to run a thing

The point of pipx is it handles the venv for you.

File: README.md [hunk 4]
```diff
@@ -70,19 +71,27 @@ Install Python 3.8 or newer with pip and tkinter somehow.
...
+porcu # or `porcupine` if you prefer
+```
 
 To easily run porcupine again later,
-go to *Settings* --> *Porcupine Settings*
+go to _Settings_ --> _Porcupine Settings_
 and check "Show Porcupine in the desktop menu system".
```
[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply README.md:82]: I don't have a Desktop Environment, so I can't test this.

File: README.md [hunk 5]
```diff
@@ -70,19 +71,27 @@ Install Python 3.8 or newer with pip and tkinter somehow.
...
+
+```bash
+pipx install --force git+https://github.com/Akuli/porcupine.git
+```
+
+This may nuke your configs, so be sure to back them up if you care about them.
+
 You can uninstall Porcupine by unchecking "Show Porcupine in the desktop menu system" in the settings
-and then deleting `porcupine-venv`.
+and then running `pipx uninstall porcupine`.
```
[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply README.md:94]: as stated above, pipx creates a venv for you automatically so it won't pollute anything

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -27,21 +28,21 @@ Then install Python 3.8 or newer and [git](https://git-scm.com/), and run these
     cd porcupine
     python3 -m venv env
     source env/bin/activate
-    pip install -r requirements.txt
-    pip install -r requirements-dev.txt
+    pip install -e ".[dev]"
```
[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply CONTRIBUTING.md:31]: I thought the whole point of this was to only manage deps in one place :sweat_smile: 

What are your concerns with "installing Porcupine to develop Porcupine"?

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -64,21 +65,24 @@ place and running `source env/bin/activate` again. You can run `deactivate` to u
...
+    type `interact` in terminal. Exit interactive mode with `ctrl + D`.
 - To see a report of test coverage, add `--cov=porcupine` to the above pytest
   command and then run `coverage html`. Open `htmlcov/index.html` in your favorite
   browser to view it.
 - `cd docs` followed by `python3 -m sphinx . build` creates HTML documentation.
   Open `docs/build/index.html` in your favorite browser to view it.
-
+- `xvfb-run pytest` (on most Linux systems) will run the tests in a headless mode.
+- `pytest --capture=fd` will prevent `print` statements and logs to print until
+  all tests are finished.
```
[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply CONTRIBUTING.md:86]: I'll be honest, it was never really clear to me why you prefer it that way, so I guess I would need an explanation of how you develop/debug without `--capture=fd` before I could explain the pros and cons of each solution

File: CONTRIBUTING.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply CONTRIBUTING.md:31]: If this is something that you think just would bother you, but doesn't need to be supported in the general case, you could do something like `pip install -r <(pip-compile --resolver=backtracking --all-extras)`

If you want this to work for everyone, run `pip-compile --resolver=backtracking --all-extras > requirements-dev.txt` as part of CI so we can manage deps in one place but still have `requirements-dev.txt` as you prefer.

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (rdbende) [reply README.md:91]: Which configs end up in the venv? Porcupine stores it's config files in `~/.config/porcupine/` on Linux, and in the equivalent location on Window.

File: README.md [hunk 8]
```diff
@@ -27,9 +28,9 @@ Most important features:
 - Multiple files can be opened at the same time like tabs in a web browser
 - The tabs can be dragged out of the window to open a new Porcupine window conveniently
 
-[Pygments]: https://pygments.org/
-[Langserver]: https://langserver.org/
-[Editorconfig]: https://editorconfig.org/
+[pygments]: https://pygments.org/
+[langserver]: https://langserver.org/
+[editorconfig]: https://editorconfig.org/
```
Comment (rdbende) [inline README.md:33]: Do these lowercased link definitions work with the capitalized references above?

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (rdbende) [reply README.md:94]: From my understanding `pipx install foo` basically does the following:

```console
python3 -m venv env
source env/bin/activate
pip install foo
```

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply README.md:91]: @rdbende you are correct, thanks for the clarification!

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply README.md:94]:
```bash
$ pipx list
venvs are in /home/bkirkbride/.local/pipx/venvs
apps are exposed on your $PATH at /home/bkirkbride/.local/bin
```

File: README.md (hunk 8 above)
Comment (benjamin-kirkbride) [reply README.md:33]: huh, apparently I have some plugin that automatically formats markdown files?

I assume it does, but I'm not 100% sure.

File: pyproject.toml [hunk 9]
```diff
@@ -26,24 +26,53 @@ license = {file = "LICENSE"}
 classifiers = ["License :: OSI Approved :: MIT License"]
 dynamic = []  # Do not attempt to import porcupine before dependencies are installed
 dependencies = [
-    "platformdirs>=3.0.0,<4.0.0",
-    "Pygments==2.12.0",
-    "toposort>=1.5",
-    "colorama>=0.2.5",
+    "platformdirs>=3.10.0,<4.0.0",
+    "pygments==2.16.1", # pygments version fixed because haxx in highlight plugin
```
[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply pyproject.toml:30]: good catch, sorry about that.

Should we leave the version bumped, and remove the comment, or leave it at the old version and leave the comment?

[the remark this answers is not in the document]
Comment (rdbende) [reply pyproject.toml:30]: There's #1392 already

[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply pyproject.toml:30]: Great call, I'll remove it from this then

File: CONTRIBUTING.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (Akuli) [reply CONTRIBUTING.md:31]: I like auto-generating requirements-dev.txt in CI :)

[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply CONTRIBUTING.md:31]: I'll give it a stab. I'm not exactly sure how adding and committing a file as part of CI works, but I'm sure I can figure it out.

[the remark this answers is not in the document]
Comment (Akuli) [reply CONTRIBUTING.md:31]: In Porcupine, all CI committing stuff is in `.github/workflows/autofix.yml`. But edits to that file aren't exactly very easy to test, so I would like it done in a separate PR that does as little other things as possible.

For example, you could just generate the file locally in this PR and create the autofix in a new PR.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply README.md:82]: It does work

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply README.md:94]: resolving this to consolidate the "should we suggest pipx?" question to one thread

File: CONTRIBUTING.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply CONTRIBUTING.md:31]: done, will automate in a new PR tracked by this issue: https://github.com/Akuli/porcupine/issues/1410

File: README.md (hunk 8 above)
Comment (benjamin-kirkbride) [reply README.md:33]: Confirmed it works.

File: CONTRIBUTING.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply CONTRIBUTING.md]: Are you happy with what I did in 8132859f1d0c207e3f22856ede61130390165890?

File: pyproject.toml (hunk 9 above)
[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply pyproject.toml]: Done

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply README.md:76]: @Akuli AFAIK this is the last thing that needs decided to move on this. Have your concerns been addressed in the other discussions?

[the remark this answers is not in the document]
Comment (Akuli) [reply README.md:76]: Switching to `pipx install` still feels like a downgrade to me. It isn't any less effort to run Porcupine this way with pipx, but you end up with a venv in a weird / non-obvious place. This means that after trying Porcupine for a few minutes, if you dislike it, you don't know what to delete to undo whatever `pipx` did. With an explicit venv creation, everything is neatly in one place and it's very obvious how to get rid of Porcupine.

However, the pipx command is useful for people who already have/use/like pipx. Maybe we could show both ways to do this?

**Comment by Akuli** [comment]
This seems to be no longer relevant.
